"""
–û—Å–Ω–æ–≤–Ω–æ–π gRPC –∫–ª–∏–µ–Ω—Ç —Å –º–æ–¥—É–ª—å–Ω–æ–π –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–æ–π
"""

import asyncio
import logging
from typing import Optional, Dict, Any, AsyncGenerator, Tuple, List
import importlib
import sys
from pathlib import Path
from datetime import datetime

import numpy as np

from integration.utils.resource_path import get_resource_path

from .types import ServerConfig, RetryConfig, HealthCheckConfig, RetryStrategy
from .retry_manager import RetryManager
from .connection_manager import ConnectionManager

# –ò–º–ø–æ—Ä—Ç –¥–ª—è –ø–æ–ª—É—á–µ–Ω–∏—è server_audio_format (–∏—Å—Ç–æ—á–Ω–∏–∫ –∏—Å—Ç–∏–Ω—ã)
try:
    from config.unified_config_loader import UnifiedConfigLoader
except ImportError:
    UnifiedConfigLoader = None

logger = logging.getLogger(__name__)


class GrpcClient:
    """–û—Å–Ω–æ–≤–Ω–æ–π gRPC –∫–ª–∏–µ–Ω—Ç —Å –º–æ–¥—É–ª—å–Ω–æ–π –∞—Ä—Ö–∏—Ç–µ–∫—Ç—É—Ä–æ–π"""
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        self.config = config or self._create_default_config()
        
        # –ú–æ–¥—É–ª—å–Ω—ã–µ –∫–æ–º–ø–æ–Ω–µ–Ω—Ç—ã
        self.connection_manager = ConnectionManager()
        self.retry_manager = RetryManager(
            RetryConfig(
                max_attempts=self.config.get('max_retry_attempts', 3),
                base_delay=self.config.get('retry_delay', 1.0),
                strategy=RetryStrategy.EXPONENTIAL  # –ò—Å–ø–æ–ª—å–∑—É–µ–º enum –≤–º–µ—Å—Ç–æ —Å—Ç—Ä–æ–∫–∏
            )
        )
        
        # –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è
        self._initialize_servers()
        self._setup_callbacks()
        
        # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —Å–µ—Ä–≤–µ—Ä –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é –∏–∑ –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏
        self._set_default_server()
    
    def _create_default_config(self) -> Dict[str, Any]:
        """–°–æ–∑–¥–∞–µ—Ç –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—é –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é –∏–∑ —Ü–µ–Ω—Ç—Ä–∞–ª–∏–∑–æ–≤–∞–Ω–Ω–æ–π —Å–∏—Å—Ç–µ–º—ã"""
        try:
            # –ó–∞–≥—Ä—É–∂–∞–µ–º –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—é –∏–∑ unified_config.yaml
            import yaml
            config_path = get_resource_path('config/unified_config.yaml')
            with open(config_path, 'r') as f:
                config = yaml.safe_load(f)
            
            grpc_data = config.get('grpc', {})
            servers_config = grpc_data.get('servers', {})
            
            # –ü—Ä–µ–æ–±—Ä–∞–∑—É–µ–º –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—é –≤ —Ñ–æ—Ä–º–∞—Ç, –æ–∂–∏–¥–∞–µ–º—ã–π GrpcClient
            servers = {}
            for server_name, server_config in servers_config.items():
                server_dict = {
                    'address': server_config.get('host', '127.0.0.1'),
                    'port': server_config.get('port', 50051),
                    'use_ssl': server_config.get('ssl', False),
                    'ssl_verify': server_config.get('ssl_verify', True),
                    'grpc_path': server_config.get('grpc_path'),
                    'use_http2': server_config.get('use_http2', True),
                    'timeout': server_config.get('timeout', grpc_data.get('connection_timeout', 30)),
                    'retry_attempts': server_config.get('retry_attempts', grpc_data.get('retry_attempts', 3)),
                    'retry_delay': server_config.get('retry_delay', grpc_data.get('retry_delay', 1.0)),
                    'keepalive': server_config.get('keepalive', True)
                }
                # DEBUG: –õ–æ–≥–∏—Ä—É–µ–º –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—é –∫–∞–∂–¥–æ–≥–æ —Å–µ—Ä–≤–µ—Ä–∞
                logger.info(f"üîå [DEBUG] Loaded server '{server_name}': {server_dict['address']}:{server_dict['port']}, ssl={server_dict['use_ssl']}, ssl_verify={server_dict['ssl_verify']}")
                servers[server_name] = server_dict
            
            return {
                'servers': servers,
                'auto_fallback': True,
                'health_check_interval': 30,
                'connection_timeout': grpc_data.get('connection_timeout', 10),
                'max_retry_attempts': grpc_data.get('retry_attempts', 3),
                'retry_strategy': 'exponential',
                'circuit_breaker_threshold': 5,
                'circuit_breaker_timeout': 60,
                'welcome_timeout_sec': 30.0
            }
            
        except Exception as e:
            logger.warning(f"‚ö†Ô∏è –ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å —Ü–µ–Ω—Ç—Ä–∞–ª–∏–∑–æ–≤–∞–Ω–Ω—É—é –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—é: {e}")
            # Fallback –∫ –º–∏–Ω–∏–º–∞–ª—å–Ω–æ–π –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏
            return {
                'servers': {
                    'local': {
                        'address': '127.0.0.1',
                        'port': 50051,
                        'use_ssl': False,
                        'timeout': 30,
                        'retry_attempts': 3,
                        'retry_delay': 1.0
                    }
                },
                'auto_fallback': True,
                'health_check_interval': 30,
                'connection_timeout': 10,
                'max_retry_attempts': 3,
                'retry_strategy': 'exponential',
                'circuit_breaker_threshold': 5,
                'circuit_breaker_timeout': 60,
                'welcome_timeout_sec': 30.0
            }
    
    def _initialize_servers(self):
        """–ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä—É–µ—Ç –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏ —Å–µ—Ä–≤–µ—Ä–æ–≤"""
        try:
            servers_config = self.config.get('servers', {})
            for name, server_config in servers_config.items():
                # DEBUG: Log what we're reading from config dict
                ssl_verify_from_config = server_config.get('ssl_verify', True)
                logger.info(f"üîå [DEBUG] _initialize_servers creating ServerConfig for '{name}': ssl_verify={ssl_verify_from_config}")

                config = ServerConfig(
                    address=server_config['address'],
                    port=server_config['port'],
                    use_ssl=server_config.get('use_ssl', False),
                    ssl_verify=ssl_verify_from_config,  # NEW: Add ssl_verify
                    grpc_path=server_config.get('grpc_path'),  # NEW: Add grpc_path
                    use_http2=server_config.get('use_http2', True),  # NEW: Add use_http2
                    timeout=server_config.get('timeout', 30),
                    retry_attempts=server_config.get('retry_attempts', 3),
                    retry_delay=server_config.get('retry_delay', 1.0),
                    max_message_size=server_config.get('max_message_size', 50 * 1024 * 1024),
                    keepalive=server_config.get('keepalive', True)  # NEW: Add keepalive
                )

                # DEBUG: Log the created ServerConfig
                logger.info(f"üîå [DEBUG] Created ServerConfig for '{name}': ssl_verify={config.ssl_verify}")

                self.connection_manager.add_server(name, config)

            logger.info(f"üåê –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω–æ {len(servers_config)} —Å–µ—Ä–≤–µ—Ä–æ–≤")
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏–∏ —Å–µ—Ä–≤–µ—Ä–æ–≤: {e}")
    
    def _setup_callbacks(self):
        """–ù–∞—Å—Ç—Ä–∞–∏–≤–∞–µ—Ç callback —Ñ—É–Ω–∫—Ü–∏–∏"""
        self.connection_manager.set_connection_callback(self._on_connection_changed)
        self.connection_manager.set_error_callback(self._on_error)
    
    def _get_server_audio_format(self) -> Dict[str, Any]:
        """
        –ü–æ–ª—É—á–∞–µ—Ç server_audio_format –∏–∑ unified_config (–∏—Å—Ç–æ—á–Ω–∏–∫ –∏—Å—Ç–∏–Ω—ã)
            
        Returns:
            Dict —Å –ø–æ–ª—è–º–∏ sample_rate, channels, dtype
        """
        # –ü–æ–ø—ã—Ç–∫–∞ 1: UnifiedConfigLoader (–ø—Ä–µ–¥–ø–æ—á—Ç–∏—Ç–µ–ª—å–Ω—ã–π —Å–ø–æ—Å–æ–±)
        if UnifiedConfigLoader is not None:
            try:
                unified_config = UnifiedConfigLoader.get_instance()
                config = unified_config._load_config()
                server_audio_format = config.get('server_audio_format', {})
                return {
                    'sample_rate': server_audio_format.get('sample_rate', 48000),
                    'channels': server_audio_format.get('channels', 1),
                    'dtype': server_audio_format.get('dtype', 'int16')
                }
            except Exception as e:
                logger.warning(f"‚ö†Ô∏è –ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å server_audio_format –∏–∑ UnifiedConfigLoader: {e}")
        
        # Fallback: –∑–∞–≥—Ä—É–∂–∞–µ–º –Ω–∞–ø—Ä—è–º—É—é –∏–∑ YAML
        try:
            import yaml
            config_path = get_resource_path('config/unified_config.yaml')
            with open(config_path, 'r') as f:
                config = yaml.safe_load(f)
                server_audio_format = config.get('server_audio_format', {})
                return {
                    'sample_rate': server_audio_format.get('sample_rate', 48000),
                    'channels': server_audio_format.get('channels', 1),
                    'dtype': server_audio_format.get('dtype', 'int16')
                }
        except Exception as e:
            logger.warning(f"‚ö†Ô∏è –ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å server_audio_format –∏–∑ YAML: {e}")
        
        # –ü–æ—Å–ª–µ–¥–Ω–∏–π fallback: –∑–Ω–∞—á–µ–Ω–∏—è –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é (Edge TTS = 48k)
            return {
            'sample_rate': 48000,
            'channels': 1,
            'dtype': 'int16'
        }
    
    def _set_default_server(self):
        """–£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ—Ç —Å–µ—Ä–≤–µ—Ä –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é –∏–∑ –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏"""
        try:
            # –¶–ï–ù–¢–†–ê–õ–ò–ó–û–í–ê–ù–û: –ò—Å–ø–æ–ª—å–∑—É–µ–º ServerManager –≤–º–µ—Å—Ç–æ –ø—Ä—è–º–æ–≥–æ —á—Ç–µ–Ω–∏—è YAML
            from config.server_manager import get_default_server
            
            default_server = get_default_server() or 'local'

            # DEBUG: –õ–æ–≥–∏—Ä—É–µ–º –≤—ã–±–æ—Ä —Å–µ—Ä–≤–µ—Ä–∞
            logger.info(f"üîå [DEBUG] Config says default server: '{default_server}'")
            logger.info(f"üîå [DEBUG] Available servers: {list(self.connection_manager.servers.keys())}")

            # –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º —Å–µ—Ä–≤–µ—Ä –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é
            if default_server in self.connection_manager.servers:
                self.connection_manager.current_server = default_server
                logger.info(f"üåê –£—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω —Å–µ—Ä–≤–µ—Ä –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é: {default_server}")
            else:
                logger.warning(f"‚ö†Ô∏è –°–µ—Ä–≤–µ—Ä '{default_server}' –Ω–µ –Ω–∞–π–¥–µ–Ω, –∏—Å–ø–æ–ª—å–∑—É–µ–º 'local'")
                self.connection_manager.current_server = 'local'

        except Exception as e:
            logger.warning(f"‚ö†Ô∏è –ù–µ —É–¥–∞–ª–æ—Å—å –∑–∞–≥—Ä—É–∑–∏—Ç—å –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏—é gRPC: {e}")
            # –ò—Å–ø–æ–ª—å–∑—É–µ–º local –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é
            self.connection_manager.current_server = 'local'
    
    def _on_connection_changed(self, state):
        """–û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç –∏–∑–º–µ–Ω–µ–Ω–∏—è —Å–æ—Å—Ç–æ—è–Ω–∏—è —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è"""
        logger.info(f"üîÑ –°–æ—Å—Ç–æ—è–Ω–∏–µ —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è: {state.value}")
    
    def _on_error(self, error: Exception, context: str):
        """–û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç –æ—à–∏–±–∫–∏"""
        logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤ {context}: {error}")
    
    async def connect(self, server_name: Optional[str] = None) -> bool:
        """–ü–æ–¥–∫–ª—é—á–∞–µ—Ç—Å—è –∫ —Å–µ—Ä–≤–µ—Ä—É"""
        return await self.connection_manager.connect(server_name)
    
    async def disconnect(self):
        """–û—Ç–∫–ª—é—á–∞–µ—Ç—Å—è –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞"""
        await self.connection_manager.disconnect()
    
    async def switch_server(self, server_name: str) -> bool:
        """–ü–µ—Ä–µ–∫–ª—é—á–∞–µ—Ç—Å—è –Ω–∞ –¥—Ä—É–≥–æ–π —Å–µ—Ä–≤–µ—Ä"""
        return await self.connection_manager.switch_server(server_name)
    
    async def execute_with_retry(self, operation, *args, **kwargs):
        """–í—ã–ø–æ–ª–Ω—è–µ—Ç –æ–ø–µ—Ä–∞—Ü–∏—é —Å retry –º–µ—Ö–∞–Ω–∏–∑–º–æ–º"""
        return await self.retry_manager.execute_with_retry(operation, *args, **kwargs)

    def get_connection_state(self):
        """–í–æ–∑–≤—Ä–∞—â–∞–µ—Ç —Ç–µ–∫—É—â–µ–µ —Å–æ—Å—Ç–æ—è–Ω–∏–µ —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è"""
        return self.connection_manager.get_connection_state()
    
    def get_metrics(self):
        """–í–æ–∑–≤—Ä–∞—â–∞–µ—Ç –º–µ—Ç—Ä–∏–∫–∏ —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è"""
        return self.connection_manager.get_metrics()
    
    def is_connected(self) -> bool:
        """–ü—Ä–æ–≤–µ—Ä—è–µ—Ç, –ø–æ–¥–∫–ª—é—á–µ–Ω –ª–∏ –∫–ª–∏–µ–Ω—Ç"""
        return self.connection_manager.is_connected()
    
    async def stream_audio(
        self,
        prompt: str,
        screenshot_base64: str,
        screen_info: dict,
        hardware_id: str,
        session_id: str,
        *,
        timeout: Optional[float] = None,
    ) -> AsyncGenerator[Any, None]:
        """–°—Ç—Ä–∏–º–∏–Ω–≥ –∞—É–¥–∏–æ –∏ —Ç–µ–∫—Å—Ç–∞ –Ω–∞ —Å–µ—Ä–≤–µ—Ä
        
        Args:
            prompt: –¢–µ–∫—Å—Ç–æ–≤–∞—è –∫–æ–º–∞–Ω–¥–∞ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—è
            screenshot_base64: Base64-–∫–æ–¥–∏—Ä–æ–≤–∞–Ω–Ω—ã–π —Å–∫—Ä–∏–Ω—à–æ—Ç
            screen_info: –ò–Ω—Ñ–æ—Ä–º–∞—Ü–∏—è –æ —Ä–∞–∑–º–µ—Ä–µ —ç–∫—Ä–∞–Ω–∞ (dict —Å 'width' –∏ 'height' –∏–ª–∏ –æ–±—ä–µ–∫—Ç —Å –∞—Ç—Ä–∏–±—É—Ç–∞–º–∏)
            hardware_id: –£–Ω–∏–∫–∞–ª—å–Ω—ã–π Hardware ID –æ–±–æ—Ä—É–¥–æ–≤–∞–Ω–∏—è
            session_id: ID —Å–µ—Å—Å–∏–∏ (–æ–±—è–∑–∞—Ç–µ–ª–µ–Ω, –µ–¥–∏–Ω—Å—Ç–≤–µ–Ω–Ω—ã–π –∏—Å—Ç–æ—á–Ω–∏–∫ –∏—Å—Ç–∏–Ω—ã - ApplicationStateManager)
        """
        # –ö–†–ò–¢–ò–ß–ù–û: Fail-fast –ø—Ä–æ–≤–µ—Ä–∫–∞ session_id
        # –ü—Ä–µ–æ–±—Ä–∞–∑—É–µ–º session_id –≤ —Å—Ç—Ä–æ–∫—É (–º–æ–∂–µ—Ç –±—ã—Ç—å float –∏–ª–∏ –¥—Ä—É–≥–æ–π —Ç–∏–ø)
        session_id = str(session_id) if session_id is not None else ""
        if not session_id or not session_id.strip():
            error_msg = "session_id is required and must be provided (Source of Truth: ApplicationStateManager)"
            logger.error(f"‚ùå [gRPC] {error_msg}")
            raise ValueError(error_msg)
        
        try:
            logger.info(f"üîç screen_info type: {type(screen_info)}")
            logger.info(f"üîç screen_info content: {screen_info}")
            
            if not self.is_connected():
                await self.connect()

            # –ò–º–ø–æ—Ä—Ç–∏—Ä—É–µ–º protobuf-–º–æ–¥—É–ª–∏ —Å —Ñ–æ–ª–±—ç–∫–æ–º –Ω–∞ server/
            streaming_pb2, streaming_pb2_grpc = self._import_proto_modules()
            
            # –°–æ–∑–¥–∞–µ–º –∑–∞–ø—Ä–æ—Å
            if hasattr(screen_info, 'get'):
                # –≠—Ç–æ —Å–ª–æ–≤–∞—Ä—å
                screen_width = screen_info.get('width')
                screen_height = screen_info.get('height')
            elif hasattr(screen_info, 'width') and hasattr(screen_info, 'height'):
                # –≠—Ç–æ –æ–±—ä–µ–∫—Ç —Å –∞—Ç—Ä–∏–±—É—Ç–∞–º–∏ width –∏ height
                screen_width = screen_info.width  # type: ignore[attr-defined]
                screen_height = screen_info.height  # type: ignore[attr-defined]
            else:
                # –ù–µ–∏–∑–≤–µ—Å—Ç–Ω—ã–π —Ç–∏–ø, –∏—Å–ø–æ–ª—å–∑—É–µ–º –∑–Ω–∞—á–µ–Ω–∏—è –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é
                logger.warning(f"‚ö†Ô∏è –ù–µ–∏–∑–≤–µ—Å—Ç–Ω—ã–π —Ç–∏–ø screen_info: {type(screen_info)}, –∏—Å–ø–æ–ª—å–∑—É–µ–º –∑–Ω–∞—á–µ–Ω–∏—è –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é")
                screen_width = None
                screen_height = None
            
            request = streaming_pb2.StreamRequest(
                prompt=prompt,
                screenshot=screenshot_base64,
                screen_width=screen_width,
                screen_height=screen_height,
                hardware_id=hardware_id,
                session_id=session_id  # –ö–†–ò–¢–ò–ß–ù–û: –ø–µ—Ä–µ–¥–∞–µ–º session_id –∏–∑ ApplicationStateManager
            )
            
            # –í—ã–ø–æ–ª–Ω—è–µ–º —Å—Ç—Ä–∏–º–∏–Ω–≥
            rpc_timeout = timeout if timeout and timeout > 0 else None
            async for response in streaming_pb2_grpc.StreamingServiceStub(
                self.connection_manager.channel
            ).StreamAudio(request, timeout=rpc_timeout):
                yield response
                
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ —Å—Ç—Ä–∏–º–∏–Ω–≥–∞ –∞—É–¥–∏–æ: {e}")
            raise

    async def generate_welcome_audio(
        self,
        text: str,
        *,
        voice: Optional[str] = None,
        language: Optional[str] = None,
        session_id: Optional[str] = None,
        timeout: Optional[float] = None,
        server_name: Optional[str] = None,
    ) -> Dict[str, Any]:
        """–ó–∞–ø—Ä–∞—à–∏–≤–∞–µ—Ç —Å–µ—Ä–≤–µ—Ä–Ω—É—é –≥–µ–Ω–µ—Ä–∞—Ü–∏—é –ø—Ä–∏–≤–µ—Ç—Å—Ç–≤–µ–Ω–Ω–æ–≥–æ –∞—É–¥–∏–æ.

        Returns dict c numpy –º–∞—Å—Å–∏–≤–æ–º –∞—É–¥–∏–æ –∏ –º–µ—Ç–∞–¥–∞–Ω–Ω—ã–º–∏.
        """
        if not text or not text.strip():
            raise ValueError("Welcome text must be non-empty")

        target_server = server_name or self.connection_manager.current_server

        if not self.is_connected():
            await self.connect(target_server)
        elif server_name and self.connection_manager.current_server != server_name:
            await self.connection_manager.switch_server(server_name)

        streaming_pb2, streaming_pb2_grpc = self._import_proto_modules()

        request = streaming_pb2.WelcomeRequest(
            text=text,
            session_id=session_id or f"welcome_{datetime.now().timestamp()}",
        )

        if voice:
            request.voice = voice
        if language:
            request.language = language

        stub = streaming_pb2_grpc.StreamingServiceStub(self.connection_manager.channel)
        rpc_timeout = timeout or self.config.get('welcome_timeout_sec', 30.0)

        audio_chunks: List[bytes] = []
        metadata: Dict[str, Any] = {}
        chunk_dtype: Optional[str] = None

        try:
            async for response in stub.GenerateWelcomeAudio(request, timeout=rpc_timeout):
                content = response.WhichOneof('content')
                if content == 'audio_chunk':
                    chunk = response.audio_chunk
                    if chunk.audio_data:
                        audio_bytes = bytes(chunk.audio_data)
                        if audio_bytes:
                            audio_chunks.append(audio_bytes)
                            chunk_dtype = chunk.dtype or chunk_dtype
                elif content == 'metadata':
                    metadata = {
                        'method': response.metadata.method,
                        'duration_sec': response.metadata.duration_sec,
                        'sample_rate': response.metadata.sample_rate,
                        'channels': response.metadata.channels,
                        'dtype': getattr(response.metadata, 'dtype', None),  # –ü—Ä–æ–≤–µ—Ä—è–µ–º dtype –∏–∑ metadata
                    }
                elif content == 'error_message':
                    raise RuntimeError(response.error_message)
                elif content == 'end_message':
                    break

        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≥–µ–Ω–µ—Ä–∞—Ü–∏–∏ –ø—Ä–∏–≤–µ—Ç—Å—Ç–≤–µ–Ω–Ω–æ–≥–æ –∞—É–¥–∏–æ: {e}")
            raise

        if not audio_chunks:
            raise RuntimeError("Server returned no audio data")

        raw_bytes = b''.join(audio_chunks)
        
        # –û–ø—Ä–µ–¥–µ–ª—è–µ–º dtype: –ø—Ä–∏–æ—Ä–∏—Ç–µ—Ç metadata.dtype > chunk_dtype > server_audio_format > fallback int16
        metadata_dtype = metadata.get('dtype')
        if metadata_dtype:
            dtype_str = str(metadata_dtype).lower()
        elif chunk_dtype:
            dtype_str = str(chunk_dtype).lower()
        else:
            # Fallback: –∏—Å–ø–æ–ª—å–∑—É–µ–º server_audio_format –∏–ª–∏ int16
            server_audio_format = self._get_server_audio_format()
            dtype_str = str(server_audio_format.get('dtype', 'int16')).lower()
        
        # –ù–æ—Ä–º–∞–ª–∏–∑—É–µ–º dtype –∏ –≤—ã–±–∏—Ä–∞–µ–º numpy —Ç–∏–ø
        if dtype_str in ('int16', 'pcm_s16le', 'short', 's16'):
            np_dtype = np.int16
            dtype_str = 'int16'
        elif dtype_str in ('float32', 'float', 'f32'):
            np_dtype = np.float32
            dtype_str = 'float32'
        elif dtype_str in ('int32', 's32'):
            np_dtype = np.int32
            dtype_str = 'int32'
        else:
            logger.warning(f"‚ö†Ô∏è –ù–µ–ø–æ–¥–¥–µ—Ä–∂–∏–≤–∞–µ–º—ã–π dtype '{dtype_str}', –ø—Ä–∏–≤–æ–∂—É –∫ int16")
            np_dtype = np.int16
            dtype_str = 'int16'
        
        audio_array = np.frombuffer(raw_bytes, dtype=np_dtype)

        # üîç –î–ò–ê–ì–ù–û–°–¢–ò–ö–ê: –§–∏–Ω–∞–ª—å–Ω—ã–µ –º–µ—Ç–∞–¥–∞–Ω–Ω—ã–µ –ø–µ—Ä–µ–¥ –≤–æ–∑–≤—Ä–∞—Ç–æ–º
        # –ü—Ä–∏–æ—Ä–∏—Ç–µ—Ç: metadata –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞ > server_audio_format (–∏—Å—Ç–æ—á–Ω–∏–∫ –∏—Å—Ç–∏–Ω—ã) > fallback
        server_audio_format = self._get_server_audio_format()
        
        # –ï—Å–ª–∏ metadata —Å–æ–¥–µ—Ä–∂–∏—Ç sample_rate/channels/dtype ‚Üí –∏—Å–ø–æ–ª—å–∑—É–µ–º –∏—Ö (–ø—Ä–∏–æ—Ä–∏—Ç–µ—Ç metadata)
        # –ï—Å–ª–∏ –Ω–µ—Ç ‚Üí –∏—Å–ø–æ–ª—å–∑—É–µ–º server_audio_format (–∏—Å—Ç–æ—á–Ω–∏–∫ –∏—Å—Ç–∏–Ω—ã)
        # dtype —É–∂–µ –æ–ø—Ä–µ–¥–µ–ª–µ–Ω –≤—ã—à–µ –ø—Ä–∏ –¥–µ–∫–æ–¥–∏—Ä–æ–≤–∞–Ω–∏–∏ (dtype_str)
        metadata_sr = metadata.get('sample_rate')
        metadata_ch = metadata.get('channels')
        
        final_sample_rate = int(metadata_sr) if metadata_sr is not None else int(server_audio_format.get('sample_rate', 48000))
        final_channels = int(metadata_ch) if metadata_ch is not None else int(server_audio_format.get('channels', 1))
        final_dtype = dtype_str  # –ò—Å–ø–æ–ª—å–∑—É–µ–º dtype, –æ–ø—Ä–µ–¥–µ–ª–µ–Ω–Ω—ã–π –ø—Ä–∏ –¥–µ–∫–æ–¥–∏—Ä–æ–≤–∞–Ω–∏–∏
        
        # –ò—Å–ø–æ–ª—å–∑—É–µ–º final_channels –¥–ª—è reshape (–±–µ–∑–æ–ø–∞—Å–Ω–æ, –¥–∞–∂–µ –µ—Å–ª–∏ metadata –ø—É—Å—Ç–∞—è)
        if final_channels > 1:
            try:
                audio_array = audio_array.reshape(-1, final_channels)
            except Exception:
                logger.warning("‚ö†Ô∏è –ù–µ —É–¥–∞–ª–æ—Å—å –∏–∑–º–µ–Ω–∏—Ç—å —Ñ–æ—Ä–º—É –∞—É–¥–∏–æ –ø–æ –∫–∞–Ω–∞–ª–∞–º, –æ—Å—Ç–∞–≤–ª—è—é –æ–¥–Ω–æ–º–µ—Ä–Ω—ã–π –º–∞—Å—Å–∏–≤")

        logger.info(
            f"üîç [GRPC_DIAG] –§–∏–Ω–∞–ª—å–Ω—ã–π sample_rate: {final_sample_rate}Hz "
            f"(–∏–∑ metadata: {metadata.get('sample_rate', 'N/A')}, "
            f"–∏–∑ server_audio_format: {server_audio_format.get('sample_rate', 'N/A')})"
            )
        
        # üîç –î–ò–ê–ì–ù–û–°–¢–ò–ö–ê: –í—ã—á–∏—Å–ª—è–µ–º —Ä–µ–∞–ª—å–Ω—É—é –¥–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å
        actual_samples = len(audio_array) if audio_array.ndim == 1 else audio_array.shape[0]
        calculated_duration = actual_samples / float(final_sample_rate) if final_sample_rate > 0 else 0.0
        metadata_duration = metadata.get('duration_sec')
        logger.info(
            f"üîç [GRPC_DIAG] –§–∏–Ω–∞–ª—å–Ω—ã–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç: samples={actual_samples}, "
            f"sr={final_sample_rate}Hz, calculated_duration={calculated_duration:.3f}s, "
            f"metadata_duration={metadata_duration if metadata_duration is not None else 'N/A'}"
        )
        
        result = {
            'audio': audio_array,
            'metadata': {
                'method': metadata.get('method', 'server'),
                'duration_sec': metadata.get('duration_sec'),
                'sample_rate': final_sample_rate,
                'channels': final_channels,
                'dtype': final_dtype,
            }
        }

        return result

    async def stream_tts_audio(
        self,
        text: str,
        *,
        voice: Optional[str] = None,
        language: Optional[str] = None,
        session_id: Optional[str] = None,
        timeout: Optional[float] = None,
    ) -> AsyncGenerator[Dict[str, Any], None]:
        """Streams TTS audio chunks via GenerateWelcomeAudio.
        
        Unlike generate_welcome_audio which buffers all audio, this yields each chunk
        immediately for real-time playback.
        
        Yields:
            Dict with 'audio_chunk' data: dtype, sample_rate, channels, bytes
        """
        if not text or not text.strip():
            logger.warning("stream_tts_audio called with empty text")
            return

        if not self.is_connected():
            await self.connect()

        streaming_pb2, streaming_pb2_grpc = self._import_proto_modules()

        request = streaming_pb2.WelcomeRequest(
            text=text,
            session_id=session_id or f"tts_{datetime.now().timestamp()}",
        )
        if voice:
            request.voice = voice
        if language:
            request.language = language

        stub = streaming_pb2_grpc.StreamingServiceStub(self.connection_manager.channel)
        rpc_timeout = timeout or self.config.get('welcome_timeout_sec', 30.0)

        try:
            async for response in stub.GenerateWelcomeAudio(request, timeout=rpc_timeout):
                content = response.WhichOneof('content')
                if content == 'audio_chunk':
                    ch = response.audio_chunk
                    if ch.audio_data and len(ch.audio_data) > 0:
                        yield {
                            'type': 'audio_chunk',
                            'bytes': bytes(ch.audio_data),
                            'dtype': ch.dtype or 'int16',
                            'sample_rate': ch.sample_rate or 48000,
                            'channels': ch.channels or 1,
                            'shape': list(ch.shape) if ch.shape else [],
                        }
                elif content == 'metadata':
                    yield {
                        'type': 'metadata',
                        'method': response.metadata.method,
                        'duration_sec': response.metadata.duration_sec,
                        'sample_rate': response.metadata.sample_rate,
                        'channels': response.metadata.channels,
                    }
                elif content == 'error_message':
                    logger.error(f"TTS error: {response.error_message}")
                    yield {'type': 'error', 'message': response.error_message}
                    break
                elif content == 'end_message':
                    yield {'type': 'end', 'message': response.end_message}
                    break
        except Exception as e:
            logger.error(f"‚ùå stream_tts_audio error: {e}")
            yield {'type': 'error', 'message': str(e)}

    async def cleanup(self):
        """–û—á–∏—Å—Ç–∫–∞ —Ä–µ—Å—É—Ä—Å–æ–≤"""
        try:
            await self.connection_manager.cleanup()
            logger.info("üßπ GrpcClient –æ—á–∏—â–µ–Ω")
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –æ—á–∏—Å—Ç–∫–∏ GrpcClient: {e}")

    def _import_proto_modules(self) -> Tuple[Any, Any]:
        """–ì–∏–±–∫–∏–π –∏–º–ø–æ—Ä—Ç streaming_pb2 –∏ streaming_pb2_grpc.
        –°–Ω–∞—á–∞–ª–∞ –ø—Ä–æ–±—É–µ–º –∏–∑ proto –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏–∏ –º–æ–¥—É–ª—è, –∑–∞—Ç–µ–º fallback –≤ server/.
        """
        # 1) –ü—ã—Ç–∞–µ–º—Å—è –∏–º–ø–æ—Ä—Ç–∏—Ä–æ–≤–∞—Ç—å –∏–∑ proto –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏–∏ –º–æ–¥—É–ª—è
        try:
            # –ü—É—Ç—å: client/modules/grpc_client/proto/
            proto_dir = Path(__file__).resolve().parent.parent / 'proto'
            
            if proto_dir.exists() and str(proto_dir) not in sys.path:
                sys.path.insert(0, str(proto_dir))
                logger.info(f"‚úÖ –î–æ–±–∞–≤–ª–µ–Ω –ø—É—Ç—å –∫ proto –º–æ–¥—É–ª—è–º: {proto_dir}")
            
            pb2 = importlib.import_module('streaming_pb2')
            pb2_grpc = importlib.import_module('streaming_pb2_grpc')
            logger.info("‚úÖ Protobuf –º–æ–¥—É–ª–∏ —É—Å–ø–µ—à–Ω–æ –∏–º–ø–æ—Ä—Ç–∏—Ä–æ–≤–∞–Ω—ã –∏–∑ proto/")
            return pb2, pb2_grpc
        except Exception as local_err:
            logger.warning(f"‚ö†Ô∏è –ù–µ —É–¥–∞–ª–æ—Å—å –∏–º–ø–æ—Ä—Ç–∏—Ä–æ–≤–∞—Ç—å –∏–∑ proto/: {local_err}")

        # 2) –ü—ã—Ç–∞–µ–º—Å—è –≤–∑—è—Ç—å –∏–∑ server/ (—Ä–µ–ø–æ–∑–∏—Ç–æ—Ä–∏–π –∫–æ—Ä–µ–Ω—å/ server)
        try:
            repo_root = Path(__file__).resolve().parents[4]
            server_dir = repo_root / 'server'
            
            # –ü—Ä–æ–≤–µ—Ä—è–µ–º —Å—É—â–µ—Å—Ç–≤–æ–≤–∞–Ω–∏–µ –∏ –¥–æ–±–∞–≤–ª—è–µ–º —Ç–æ–ª—å–∫–æ –µ—Å–ª–∏ –Ω—É–∂–Ω–æ
            if server_dir.exists() and str(server_dir) not in sys.path:
                sys.path.append(str(server_dir))
                logger.info(f"‚úÖ –î–æ–±–∞–≤–ª–µ–Ω –ø—É—Ç—å –∫ server –º–æ–¥—É–ª—è–º: {server_dir}")
            
            pb2 = importlib.import_module('streaming_pb2')
            pb2_grpc = importlib.import_module('streaming_pb2_grpc')
            logger.info("‚úÖ Protobuf –º–æ–¥—É–ª–∏ —É—Å–ø–µ—à–Ω–æ –∏–º–ø–æ—Ä—Ç–∏—Ä–æ–≤–∞–Ω—ã –∏–∑ server/")
            return pb2, pb2_grpc
        except Exception as e:
            raise ImportError(f"Unable to import protobuf modules (streaming_pb2*). Error: {e}")
